plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java

```diff
@@ -1653,6 +1657,18 @@ public void close()
             }
         }
 
+        private void ensureExtraBytesCapacity()
+        {
+            if (buffer.length < INITIAL_BUFFER_SIZE) {
+                buffer = Arrays.copyOf(buffer, INITIAL_BUFFER_SIZE);
+            }
+            else if (buffer.length < partSize) {
+                // grow array by 50%
+                int newBytesLength = min(buffer.length + (buffer.length >> 1), partSize);
+                buffer = Arrays.copyOf(buffer, newBytesLength);
```

Comment:
You've removed total capacity part. Now buffer (after increase) might be smaller than what you want to put inside.

The previous code was correct, it just required capping at `partSize`